File: pbzx.rb
require"xz";i=File.new($*[0],"rb");o=File.new($*[1],"wb");i.read(12)
while b=i.read(8)
b=b.unpack("Q>")[0];c=i.read(8).unpack("Q>")[0];e=i.read(c)
o.write(b>c ?XZ.decompress(e):e)
end

File: pbzx.py
import lzma,struct,sys;x,y,z=sys.argv;u=struct.unpack;r=open(y,'rb').read;w=open(z,'wb').write;r(12);b=r(16)
while b:
 l,s=u('>QQ',b);d=r(s)
 if l>s:d=lzma.decompress(d)
 w(d);b=r(16)